### PR TITLE
Fusion Find

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
This PR adds find functionality from convolution code to fusion.

Progress:
1. Untangle find-related code from convolution: done
2. Reuse core find-related code: done.
3. Add find-db support: done except for find modes.

Future work:
1. Find modes support: this is extremely cheap to add after find db and allows to use hybrid mode and such. Can be done in a this PR if needed, but I would rather split it out.

Some notes:
It is impossible to completely unify the code between API call and solver method among primitives due to API difference outside of find 2.0, which was designed with that in mind.
This would be optimized a bit after #2146 gets merged due to the changes to how find works there.

### CODE REVIEW

**Comment by DrizztDoUrden** [comment]
I think I would split removal of convolution context to a separate branch and then rebase this one
That would limit the time I would have to spend maintaining a pr with merge conflicts

**Comment by DrizztDoUrden** [comment]
This still needs a bit more work.

**Comment by DrizztDoUrden** [comment]
@JehandadKhan @averinevg @junliume it's ready for review

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -249,7 +249,8 @@ set( MIOpen_Source
     tensor.cpp
     tensor_api.cpp
     seq_tensor.cpp
-    )
+    fusion/problem_description.cpp
```
Comment (averinevg) [inline src/CMakeLists.txt]: It is better to arrange file paths in alphabetical order. And this also applies to `conv/solver_finders.cpp`.

File: src/fusion.cpp [hunk 2]
```diff
@@ -542,137 +683,144 @@ std::string FusionPlanDescriptor::GetAlgorithmName(const Handle& /*handle*/)
...
-    {
-        MIOPEN_LOG_I("No supported fusion solvers found during Search Mode.");
-        return miopenStatusUnsupportedOp;
-    }
-    // tmp_sols is a collection of ConvSolutions that isApplicable for the fusion_problem.
-    // These ConvSolutions stores instructions on how to build. It also stores invoker.
-    const auto tmp_sols = solvers.SearchForAllSolutions(
-        fusion_ctx, fusion_problem, miopen::GetDb(fusion_ctx), invoke_params);
+
+    // tmp_sols is a collection of ConvSolutions that have been return from Find for the
```
Comment (averinevg) [inline src/fusion.cpp]:
```suggestion
    // sols is a collection of ConvSolutions that have been returned from Find for the
```

File: src/fusion/problem_description.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (averinevg) [inline src/fusion/problem_description.cpp]:
```suggestion
 * Copyright (c) 2023 Advanced Micro Devices, Inc.
```

**Review by CAHEK7 — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Comment by DrizztDoUrden** [comment]
@JehandadKhan ping

**Comment by DrizztDoUrden** [comment]
@JehandadKhan ping

**Pull request merged**